### ISSUE TITLE
Wrong handling of writer after returning BINSON_ID_BUF_FULL

### ISSUE THREAD

**Issue by SylwBar** [assaabloy-ppi/binson-c-light#11]
There are two problems with writer state after returning BINSON_ID_BUF_FULL.

Following program exposes problem:

```
#include <stdio.h>
#include "binson_light.h"

int main()
{

   uint8_t buf[4];
   binson_writer w;

   binson_writer_init(&w, buf, sizeof(buf));
   binson_write_object_begin(&w);
   printf("Counter: %d, Status: %d\n", binson_writer_get_counter(&w), w.error_flags);

   binson_write_integer(&w, 1000000);
   printf("Counter: %d, Status: %d\n", binson_writer_get_counter(&w), w.error_flags);

   binson_write_integer(&w, 1);
   printf("Counter: %d, Status: %d\n", binson_writer_get_counter(&w), w.error_flags);

   return 0;
}

```
Result is:
Counter: 1, Status: 0
Counter: 6, Status: 240
Counter: 8, Status: 240

**Problem1:**
binson_write_integer(&w, 1000000) call returns BINSON_ID_BUF_FULL (240) which is correct. However counter is increased. I expect that counter shouldn't be updated in this case.
_**Root cause:**_ 
Function _binson_io_write():
line io->counter += sz; should be located after two if's

**Problem2: (after fixing problem 1)**
binson_write_integer(&w, 1) call returns BINSON_ID_BUF_FULL (240) which is incorrect. This operation should return BINSON_ID_OK as there is enough place in buffer.
_**Root cause:**_ 
Function _binson_writer_write_token():
Section
if (res)
    pwriter->error_flags = res;

is preventing error_flags update when result is BINSON_ID_OK, in this case it return previous status BINSON_ID_BUF_FULL.
There should be no if(res) - status should be updated every time.

**Comment by ppmag** [comment]
Sylwester, thank you for so deep issue research.
Fixed in v2 branch now.

**Comment by ppmag** [comment]
I forgot the reason I implemented it this way.

This behavior is expected and "as designed":
```
Counter: 1, Status: 0
Counter: 6, Status: 240
Counter: 8, Status: 240
```
Counter counts not bytes written but bytes requested to write.
Originally it was required for binson size calculations,
by issuing write commands but without any memory updates at all.
For exampe use of it pls see:
https://github.com/assaabloy-ppi/binson-c-light/blob/v2/binson_light_examples.c#L40

In case you know better way to do on-the-fly binson raw bytes size calculations pls let me know.

Best,
Alex

**Comment by ppmag** [comment]
Forgot to say, 
to force all writes go to /dev/null 
all what is required is to specify NULL buffer.
https://github.com/assaabloy-ppi/binson-c-light/blob/v2/binson_light_examples.c#L99

PS: maybe this should be better documented, I suppose...

**Comment by SylwBar** [comment]
Thanks for clarification. 
I was not aware that you're using binson_writer this way.
I agree that changes made should be revoked, I'll update my model to match this requirement.

**Comment by SylwBar** [comment]
I need to reopen this.
I've rechecked solution using initial example, and current output is:

Counter: 1, Status: 0
Counter: 6, Status: 240
Counter: 8, Status: 0

The counter works as expected, but status is cleared, despite buffer is overflowed.

I believe this is wrong, because information about buffer overflow is lost.
I can imagine situation when someone is writing data in a batch and only checks status at the end.

**Comment by SylwBar** [comment]
Looks like not required "fix" for #11 was not completely removed. Updated code.
